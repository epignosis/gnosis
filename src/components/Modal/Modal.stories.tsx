import React, { useState } from "react";
import { Story } from "@storybook/react";
import Modal, { ReactModalProps } from "./Modal";
import { Button, Input } from "@components";

export default {
  component: Modal,
  title: "Components/Modal",
  parameters: {
    controls: { hideNoControlsWarning: true },
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg", "fullscreen"],
      },
    },
  },
  args: {
    size: "md",
  },
};

export const Default: Story<ReactModalProps> = (args) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button onClick={(): void => setIsOpen(true)}>Simple Modal</Button>
      <Modal {...args} isOpen={isOpen} onClose={(): void => setIsOpen(false)}>
        <Modal.Header>This is the modal title</Modal.Header>
        <Modal.Body>This is the modal body</Modal.Body>
        <Modal.Footer>This is the modal footer</Modal.Footer>
      </Modal>
    </div>
  );
};

export const ModalWithNoHeader: Story<ReactModalProps> = (args) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button onClick={(): void => setIsOpen(true)}>Modal with no Header</Button>
      <Modal {...args} isOpen={isOpen} onClose={(): void => setIsOpen(false)}>
        <Modal.Body style={{ padding: 0 }}>
          <div style={{ overflow: "auto", maxHeight: "32rem" }}>
            <img
              src="https://res.cloudinary.com/tsevdos/image/upload/v1607431472/Card01_Sales_rtxuhj.jpg"
              style={{ width: "100%" }}
            />
          </div>
        </Modal.Body>
        <Modal.Footer>
          <div style={{ display: "flex", justifyContent: "center" }}>
            <Button onClick={(): void => setIsOpen(false)}>Close Modal</Button>
          </div>
        </Modal.Footer>
      </Modal>
    </div>
  );
};

export const ModalWithCotent: Story<ReactModalProps> = (args) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button onClick={(): void => setIsOpen(true)}>Modal with Content</Button>
      <Modal {...args} isOpen={isOpen} onClose={(): void => setIsOpen(false)}>
        <Modal.Header>This is the modal title</Modal.Header>
        <Modal.Body>
          <Input
            id="password"
            type="password"
            label="Enter current password "
            style={{ marginBottom: 16 }}
          />
          <Input
            id="new-password"
            type="password"
            label="Type new password"
            style={{ marginBottom: 16 }}
          />
          <Input id="retype-psw" type="password" label="Retype password" />
        </Modal.Body>
        <Modal.Footer style={{ textAlign: "right" }}>
          <Button
            color="secondary"
            onClick={(): void => setIsOpen(false)}
            style={{ marginRight: "1rem" }}
          >
            Cancel
          </Button>
          <Button onClick={(): void => setIsOpen(false)}>Change password</Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
};
