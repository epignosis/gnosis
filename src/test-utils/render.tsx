import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
