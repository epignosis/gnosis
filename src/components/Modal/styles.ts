import { css, Theme, SerializedStyles } from "@emotion/react";
import { Size } from "./Modal";
import { mq } from "@theme/utils/breakpoints";

const widthDimensions = {
  md: "32rem",
  lg: "64rem",
  fullscreen: "100%",
};

export const modalContent = css`
  padding: 1.5rem;
`;

export const modalHeader = ({ modal }: Theme): SerializedStyles => {
  return css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 1.25rem;
    border-bottom: 1px solid ${modal.border};

    .close-btn {
      svg {
        color: ${modal.color};

        &:hover {
          color: ${modal.iconHover};
        }
      }
    }
  `;
};

export const modalFooter = ({ modal }: Theme): SerializedStyles => {
  return css`
    padding: 1.25rem;
    border-top: 1px solid ${modal.border};
  `;
};

// plain styles as string
export const portalStyles = (size: Size): SerializedStyles => {
  return css`
    .overlay-base {
      position: fixed;
      top: 0;
      bottom: 0;
      right: 0;
      left: 0;
      display: flex;
      justify-content: center;
      align-items: center;
      background-color: rgba(0, 0, 0, 0.45);
      opacity: 0;
      z-index: 9999;
      transition: opacity 0.2s ease-in-out;

      &.overlay-before {
        opacity: 0;
      }
    }

    .overlay-after {
      opacity: 1;
    }

    .content-after {
      background-color: #fff;
      height: 100%;
      width: 100%;
      border-radius: 0;

      ${mq["md"]} {
        height: auto;
        border-radius: 5px;
        width: ${widthDimensions[size]};
        max-width: 90%;
      }

      ${size === "fullscreen" &&
      `
      display: flex;
      flex-direction: column;
      justify-content: space-between;
    `}
    }
  `;
};
