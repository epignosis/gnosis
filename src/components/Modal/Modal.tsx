import React, { CSSProperties, FC, ReactElement, cloneElement, Children } from "react";
import { ClassNames } from "@emotion/react";
import ReactModal, { Props } from "react-modal";
import { modalContent, modalFooter, modalHeader, portalStyles } from "./styles";
import { CloseCircledSVG } from "@icons/core";
import { Heading } from "@components";

export type HeaderProps = {
  onClose?: () => void;
};

const Header: FC<HeaderProps> = ({ children, onClose }) => {
  const title = typeof children === "string" ? <Heading size="md">{children}</Heading> : children;
  return (
    <header css={modalHeader} data-testid="modal-header">
      <div>{title}</div>
      {onClose && (
        <a role="button" className="close-btn" aria-label="Close button" onClick={onClose}>
          <CloseCircledSVG height={28} />
        </a>
      )}
    </header>
  );
};

type BodyProps = {
  style?: CSSProperties;
};

const Body: FC<BodyProps> = ({ children, style }) => {
  return (
    <article style={style} css={modalContent}>
      {children}
    </article>
  );
};

type FooterProps = {
  style?: CSSProperties;
};

const Footer: FC<FooterProps> = ({ children, style }) => {
  return (
    <footer style={style} css={modalFooter} data-testid="modal-footer">
      {children}
    </footer>
  );
};

type ModalCompoundProps = {
  Header: FC<Omit<HeaderProps, "onClose">>;
  Body: FC<BodyProps>;
  Footer: FC<FooterProps>;
};

export type Size = "md" | "lg" | "fullscreen";

export type ReactModalProps = Pick<Props, "isOpen"> & {
  onClose: () => void;
  size?: Size;
};

const Modal: FC<ReactModalProps> & ModalCompoundProps = ({
  children,
  isOpen,
  onClose,
  size = "md",
}) => {
  const clonedChildren = Children.map(children, (child) =>
    cloneElement(child as ReactElement, {
      onClose,
    }),
  );

  return (
    <ClassNames>
      {({ css }): JSX.Element => (
        <ReactModal
          isOpen={isOpen}
          appElement={document.getElementById("app") as HTMLElement}
          onRequestClose={onClose}
          contentLabel="modal"
          overlayClassName={{
            base: "overlay-base",
            afterOpen: "overlay-after",
            beforeClose: "overlay-before",
          }}
          className={{
            base: "content-base",
            afterOpen: "content-after",
            beforeClose: "content-before",
          }}
          closeTimeoutMS={200}
          portalClassName={css(portalStyles(size))}
          ariaHideApp={false}
        >
          {clonedChildren}
        </ReactModal>
      )}
    </ClassNames>
  );
};

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

export default Modal;
