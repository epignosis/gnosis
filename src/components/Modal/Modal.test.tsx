import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Modal from "./Modal";
import { render, screen } from "@test-utils/render";

const mockedOnClose = jest.fn();

describe("<Modal>", () => {
  const headerTxt = faker.lorem.word();
  const bodyTxt = faker.lorem.sentence();
  const footerTxt = faker.lorem.words();

  it("renders correctly with Header, Body and Footer", () => {
    render(
      <Modal isOpen onClose={mockedOnClose}>
        <Modal.Header>{headerTxt}</Modal.Header>
        <Modal.Body>{bodyTxt}</Modal.Body>
        <Modal.Footer>{footerTxt}</Modal.Footer>
      </Modal>,
    );
    const header = screen.getByText(headerTxt);
    const body = screen.getByText(bodyTxt);
    const footer = screen.getByText(footerTxt);

    expect(header).toHaveTextContent(headerTxt);
    expect(body).toHaveTextContent(bodyTxt);
    expect(footer).toHaveTextContent(footerTxt);
  });

  it("renders correctly without Footer", () => {
    render(
      <Modal isOpen onClose={mockedOnClose}>
        <Modal.Header>{headerTxt}</Modal.Header>
        <Modal.Body>{bodyTxt}</Modal.Body>
      </Modal>,
    );
    const header = screen.getByText(headerTxt);
    const body = screen.getByText(bodyTxt);
    const footer = screen.queryByTestId("modal-footer");

    expect(header).toHaveTextContent(headerTxt);
    expect(body).toHaveTextContent(bodyTxt);
    expect(footer).not.toBeInTheDocument();
  });

  it("renders correctly without Header and Footer", () => {
    render(
      <Modal isOpen onClose={mockedOnClose}>
        <Modal.Body>{bodyTxt}</Modal.Body>
      </Modal>,
    );
    const body = screen.getByText(bodyTxt);
    const header = screen.queryByTestId("modal-header");
    const footer = screen.queryByTestId("modal-footer");

    expect(body).toHaveTextContent(bodyTxt);
    expect(header).not.toBeInTheDocument();
    expect(footer).not.toBeInTheDocument();
  });

  it("Header renders correctly with close btn", () => {
    render(
      <Modal isOpen onClose={mockedOnClose}>
        <Modal.Header>{headerTxt}</Modal.Header>
      </Modal>,
    );
    const closeBtn = screen.getByRole("button");

    userEvent.click(closeBtn);

    expect(mockedOnClose).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Modal isOpen onClose={mockedOnClose}>
        <Modal.Header>Test header</Modal.Header>
        <Modal.Body>Test body</Modal.Body>
        <Modal.Footer>Test footer</Modal.Footer>
      </Modal>,
    );

    expect(container).toMatchSnapshot();
  });
});
