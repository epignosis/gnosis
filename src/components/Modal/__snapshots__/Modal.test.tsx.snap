// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal> matches snapshot 1`] = `
<div
  aria-label="modal"
  aria-modal="true"
  class="content-base"
  role="dialog"
  tabindex="-1"
>
  <header
    class="css-fvsr8l-modalHeader-modalHeader"
    data-testid="modal-header"
  >
    <div>
      <h2
        class="css-1iyd91n-headingContainer-headingContainer"
      >
        Test header
      </h2>
    </div>
    <a
      aria-label="Close button"
      class="close-btn"
      role="button"
    >
      <svg-file-stub
        height="28"
      />
    </a>
  </header>
  <article
    class="css-18mo2pj-modalContent"
  >
    Test body
  </article>
  <footer
    class="css-j18ln3-modalFooter-modalFooter"
    data-testid="modal-footer"
  >
    Test footer
  </footer>
</div>
`;

exports[`<Modal> matches snapshot with cutom HTML classes 1`] = `
<div
  aria-label="modal"
  aria-modal="true"
  class="content-base"
  role="dialog"
  tabindex="-1"
>
  <header
    class="css-fvsr8l-modalHeader-modalHeader"
    data-testid="modal-header"
  >
    <div>
      <h2
        class="css-1iyd91n-headingContainer-headingContainer"
      >
        Test header
      </h2>
    </div>
    <a
      aria-label="Close button"
      class="close-btn"
      role="button"
    >
      <svg-file-stub
        height="28"
      />
    </a>
  </header>
  <article
    class="css-18mo2pj-modalContent"
  >
    Test body
  </article>
  <footer
    class="css-j18ln3-modalFooter-modalFooter"
    data-testid="modal-footer"
  >
    Test footer
  </footer>
</div>
`;
